src/System/input_enabler.rs

```diff
@@ -24,46 +24,40 @@ pub fn make_input_enabled(component: &mut component::Component, inputs: &[String
 
         for input in inputs {
             let input_edges = component.get_next_edges(location, input, component::SyncType::Input);
-            let mut zones_federation = Federation::empty(location_inv_zone.get_dimensions());
+            let mut zones_federation = Federation::empty(dimension);
 
             for edge in input_edges {
-                let mut guard_zone = location_inv_zone.clone();
-                let has_inv = if let Some(target_invariant) = component
+                let mut guard_zone = Federation::full(dimension); //location_inv_zone.clone();
+                if let Some(target_invariant) = component
                     .get_location_by_name(edge.get_target_location())
                     .get_invariant()
                 {
                     constraint_applyer::apply_constraints_to_state(
                         target_invariant,
                         component.get_declarations(),
                         &mut guard_zone,
-                    )
-                } else {
-                    false
-                };
+                    );
+                }
 
-                if edge.get_update().is_some() {
-                    let update_clocks = edge.get_update_clocks();
-                    for clock in update_clocks {
-                        let clock_index = component.get_declarations().clocks.get(clock).unwrap();
-                        guard_zone.free_clock(*clock_index);
+                if let Some(updates) = edge.get_update() {
+                    for update in updates {
+                        let cu = update.compiled(component.get_declarations());
+                        cu.apply_as_guard(&mut guard_zone);
+                        cu.apply_as_free(&mut guard_zone);
```

Comment:
It does not, let's say you have an invariant `x=y`, the update is `x:=2`, then apply_as_guard sets `x=y=2`, then freeing x becomes `y=2` whereas if the free is applied to `x=y` without apply_as_guard it would become `true`, e.g. no guard at all which is incorrect.
The idea is to get the zone in the source location where if x were set to 2 (in this example) it would satisfy the invariant in the target location, essentially we are doing the inverse of applying the update, and that is not always correct without apply_as_guard.